esp-hal-common/src/ledc/esp32/timer.rs

```diff
@@ -0,0 +1,355 @@
+    /// Return the frequency of the timer
+    fn get_freq(&self) -> Option<_fugit_MegahertzU32>;
+
+    /// Configure the timer
+    fn configure(&mut self, config: config::Config<S::ClockSourceType>) -> Result<(), Error>;
+
+    /// Check if the timer has been configured
+    fn is_configured(&self) -> bool;
+
+    /// Return the duty resolution of the timer
+    fn get_duty(&self) -> Option<config::Duty>;
+
+    /// Return the timer number
+    fn get_number(&self) -> Number;
+}
+
+/// Interface for HW configuration of timer
+pub trait TimerHW<S: TimerSpeed> {
+    /// Get the current source timer frequency from the HW
+    fn get_freq_hw(&self) -> Option<_fugit_MegahertzU32>;
+
+    /// Configure the HW for the timer
+    fn configure_hw(&self, divisor: u32);
+
+    /// Update the timer in HW
+    fn update_hw(&self);
+}
+
+/// Timer struct
+pub struct Timer<'a, S: TimerSpeed> {
+    ledc: &'a crate::pac::ledc::RegisterBlock,
+    clock_control_config: &'a Clocks,
+    number: Number,
+    duty: Option<config::Duty>,
+    configured: bool,
+    clock_source: Option<S::ClockSourceType>,
+}
+
+impl<'a, S: TimerSpeed> TimerIFace<S> for Timer<'a, S>
+where
+    Timer<'a, S>: TimerHW<S>,
+{
+    /// Return the frequency of the timer
+    fn get_freq(&self) -> Option<_fugit_MegahertzU32> {
+        self.get_freq_hw()
+    }
+
+    /// Configure the timer
+    fn configure(&mut self, config: config::Config<S::ClockSourceType>) -> Result<(), Error> {
+        self.duty = Some(config.duty);
+        self.clock_source = Some(config.clock_source);
+
+        let src_freq: u32 = self.get_freq().unwrap().to_Hz();
+        let precision = 2_u64.pow(config.duty as u32);
+        let frequency: u32 = config.frequency.raw();
+
+        let divisor = (((src_freq as u64) << 8) + ((frequency as u64 * precision) / 2))
+            / (frequency as u64 * precision);
+
+        if divisor >= 0x10_0000 || divisor == 0 {
+            return Err(Error::Divisor);
+        }
+
+        self.configure_hw(divisor as u32);
+        self.update_hw();
+
+        self.configured = true;
+
+        Ok(())
+    }
+
+    /// Check if the timer has been configured
+    fn is_configured(&self) -> bool {
+        self.configured
+    }
+
+    /// Return the duty resolution of the timer
+    fn get_duty(&self) -> Option<config::Duty> {
+        self.duty
+    }
+
+    /// Return the timer number
+    fn get_number(&self) -> Number {
+        self.number
+    }
+}
+
+impl<'a, S: TimerSpeed> Timer<'a, S> {
+    /// Create a new intance of a timer
+    pub fn new(
+        ledc: &'a ledc::RegisterBlock,
+        clock_control_config: &'a Clocks,
+        number: Number,
+    ) -> Self {
+        Timer {
+            ledc,
+            clock_control_config,
+            number,
+            duty: None,
+            configured: false,
+            clock_source: None,
+        }
+    }
+
+    // TODO This function will be relevant when we add other clk sources
```

Comment:
Will this method be required, or can it be removed?

Reply:
If we are OK with `APBClk` we can remove it, if it makes sense to also include other clock sources we should keep it.

Reply:
Let's leave it in as a TODO.